packages/grafana-data/src/types/pluginExtensions.ts

```diff
@@ -1,70 +1,74 @@
+// --------------------------------------------------------
+
+export type PluginExtensionLinkConfig<Context extends object = object> = Pick<
+  PluginExtensionLink,
+  'title' | 'description' | 'path'
+> & {
+  // The unique name of the placement
+  // Core Grafana placements are available in the `PluginExtensionPlacements` enum
+  placement: string;
+
+  // (Optional) A function that can be used to configure the extension dynamically based on the placement's context
+  configure?: (context?: Context) => Pick<PluginExtensionLink, 'title' | 'description' | 'path'> | undefined;
 };
 
-export function isPluginExtensionLink(extension: PluginExtension | undefined): extension is PluginExtensionLink {
-  if (!extension) {
-    return false;
-  }
-  return extension.type === PluginExtensionTypes.link && 'path' in extension;
-}
+export type PluginExtensionEventHelpers = {
+  // Opens a modal dialog and renders the provided React component inside it
+  openModal: (options: {
+    // The title of the modal
+    title: string;
+    // A React element that will be rendered inside the modal
+    body: React.ElementType<{ onDismiss?: () => void }>;
+  }) => void;
+};
 
-export function assertPluginExtensionLink(
-  extension: PluginExtension | undefined
-): asserts extension is PluginExtensionLink {
-  if (!isPluginExtensionLink(extension)) {
-    throw new Error(`extension is not a link extension`);
-  }
-}
+// Placements & Contexts
+// --------------------------------------------------------
 
-export function isPluginExtensionCommand(extension: PluginExtension | undefined): extension is PluginExtensionCommand {
-  if (!extension) {
-    return false;
-  }
-  return extension.type === PluginExtensionTypes.command;
+// Placements available in core Grafana
+export enum PluginExtensionPlacements {
+  DashboardPanelMenu = 'grafana/dashboard/panel/menu',
 }
 
-export function assertPluginExtensionCommand(
-  extension: PluginExtension | undefined
-): asserts extension is PluginExtensionCommand {
-  if (!isPluginExtensionCommand(extension)) {
-    throw new Error(`extension is not a command extension`);
-  }
-}
+export type PluginExtensionPanelContext = Readonly<{
+  pluginId: string;
+  id: number;
+  title: string;
+  timeRange: Readonly<RawTimeRange>;
+  timeZone: TimeZone;
+  dashboard: Readonly<Dashboard>;
+  targets: Readonly<Array<Readonly<Target>>>;
+}>;
 
-export function extensionLinkConfigIsValid(props: {
-  path?: string;
-  description?: string;
-  title?: string;
-  placement?: string;
-}) {
-  const valuesAreStrings = Object.values(props).every((val) => typeof val === 'string' && val.length);
-  const placementIsValid = props.placement?.startsWith('grafana/') || props.placement?.startsWith('plugins/');
-  return valuesAreStrings && placementIsValid;
-}
+type Dashboard = {
+  uid: string;
+  title: string;
+  tags: Readonly<Array<Readonly<string>>>;
+};
+
+type Target = {
+  pluginId: string;
+  refId: string;
+};
```

Comment:
I think we can remove the `Readonly` types here since we are doing the deepFreeze in the `getPluginExtensions`